### PR TITLE
Update create contract tests

### PR DESCRIPTION
**Pull request by sisuresh**
Related to https://github.com/stellar/rs-stellar-contract-sdk/pull/157.

### CODE REVIEW

File: stellar-contract-env-host/src/test.rs [hunk 1]
```diff
@@ -638,27 +629,206 @@ fn create_contract_test() -> Result<(), HostError> {
...
+
+    if bin.as_slice() != hash_copy.0.as_slice() {
+        panic!("return value doesn't match")
+    }
+
+    host.visit_storage(|s: &mut Storage| {
+        assert!(s.has(&storage_key)?);
+
+        match s.get(&storage_key)?.data {
+            LedgerEntryData::ContractData(ContractDataEntry {
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply stellar-contract-env-host/src/test.rs:641]: I wasn't sure how else to pull out `val`.

File: stellar-contract-env-host/src/test.rs [hunk 2]
```diff
@@ -638,27 +629,206 @@ fn create_contract_test() -> Result<(), HostError> {
...
-  $ xxd -i target/wasm32-unknown-unknown/release/example_add_i32.wasm
+$ make build
+$ xxd -i target-tiny/wasm32-unknown-unknown/release/example_create_contract.wasm
 */
 #[cfg(feature = "vm")]
 #[test]
+fn create_contract_using_parent_id_test() {
+    use sha2::{Digest, Sha256};
+    //This contract creates another contract
+    let code: [u8; 322] = [
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply stellar-contract-env-host/src/test.rs:707]: Ah no I wasn't aware of wasm-strip. I'll make sure to update the WASM blobs.

File: stellar-contract-env-host/src/test.rs [hunk 3]
```diff
@@ -638,27 +629,206 @@ fn create_contract_test() -> Result<(), HostError> {
...
+    let pre_image = xdr::HashIdPreimage::ContractIdFromEd25519(xdr::HashIdPreimageContractId {
+        ed25519: xdr::Uint256(pub_bytes.as_slice().try_into().unwrap()),
+        salt: xdr::Uint256(salt_bytes.as_slice().try_into().unwrap()),
+    });
+    let mut buf = Vec::new();
+    pre_image
+        .write_xdr(&mut buf)
+        .expect("preimage write failed");
+
+    let parent_id = xdr::Hash(Sha256::digest(buf).try_into().expect("invalid hash"));
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply stellar-contract-env-host/src/test.rs]: Added a `sha256_hash_id_preimage` function.

File: stellar-contract-env-host/src/test.rs [hunk 4]
```diff
@@ -638,27 +629,206 @@ fn create_contract_test() -> Result<(), HostError> {
...
+    .unwrap();
+    let args = host.to_host_obj(&ScObject::Vec(scvec0)).unwrap();
+
+    let p_id_sobj = ScObject::Binary(parent_id.0.try_into().unwrap());
+    let p_id_obj = host.to_host_obj(&p_id_sobj).unwrap();
+    host.call(p_id_obj.to_object(), sym.into(), args.into())
+        .unwrap();
+
+    //Validate child contract exists and code is what we expected
+    host.visit_storage(|s: &mut Storage| {
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply stellar-contract-env-host/src/test.rs]: Added a `check_new_code` function.

File: stellar-contract-env-host/src/test.rs [hunk 5]
```diff
@@ -638,27 +629,206 @@ fn create_contract_test() -> Result<(), HostError> {
...
+    let bad_pub: &[u8] = b"3d4017c3e843895a92b70aa74d1b7ebc9c982ccf2ec4968cc0cd55f12af4660c";
+    assert!(create_contract_test_helper(secret_key, bad_pub, salt, code, sig_preimage).is_err());
     Ok(())
 }
 
 /// VM test
 /**
  This is an example WASM from the SDK that unpacks two SCV_I32 arguments, adds
- them with an overflow check, and re-packs them as an SCV_I32 if successful.
+them with an overflow check, and re-packs them as an SCV_I32 if successful.
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply stellar-contract-env-host/src/test.rs]: I updated my test and the add test. The other ones don't exist in the sdk so I'll open an issue.

File: stellar-contract-env-host/src/test.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jonjove) [reply stellar-contract-env-host/src/test.rs]: Two options:

1. (This is what I would do in this context) Change it to
```rust
            LedgerEntryData::ContractData(cde) => {
                assert_eq!(cde.val, code);
            }
```

2. (This is a good option when you want to get some things from a destructure but not all of them) Change it to
```
            LedgerEntryData::ContractData(ContractDataEntry {
                val, ..
            }) => {
                assert_eq!(val, code);
            }
```
See https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#ignoring-remaining-parts-of-a-value-with- for more information.

[the remark this answers is not in the document]
Comment (jonjove) [reply stellar-contract-env-host/src/test.rs]: Actually, in either case you can also remove the brackets around the assert.

File: stellar-contract-env-host/src/test.rs [hunk 6]
```diff
@@ -638,27 +629,206 @@ fn create_contract_test() -> Result<(), HostError> {
...
+}
+
+#[test]
+fn create_contract_test() -> Result<(), HostError> {
+    let secret_key: &[u8] = b"9d61b19deffd5a60ba844af492ec2cc44449c5697b326919703bac031cae7f60";
+    let public_key: &[u8] = b"d75a980182b10ab7d54bfed3c964073a0ee172f3daa62325af021a68f707511a";
+    let salt: &[u8] = b"3d4017c3e843895a92b70aa74d1b7ebc9c982ccf2ec4968cc0cd55f12af4660c";
+    let code: &[u8] = b"70aa74d1b7ebc9c982ccf2ec4968cc0cd55f12af4";
+
+    let separator =
```
[the remark this answers is not in the document]
Comment (jonjove) [reply stellar-contract-env-host/src/test.rs:680]: We should study this more deeply in general and make sure the message we're signing is sane. Opened https://github.com/stellar/rs-stellar-contract-env/issues/153 to track this.

**Pull request merged**